File: 2755-extra-characters-in-a-string/extra-characters-in-a-string.java
class Solution {
    Set<String> set;
    String s;
    int[] dp;

    public int minExtraChar(String s, String[] dictionary) {
        this.s = s;
        set = new HashSet<>(Arrays.asList(dictionary));
        dp = new int[s.length()];
        Arrays.fill(dp, -1);
        return dfs(0);
    }

    private int dfs(int i) {
        if (i == s.length()) return 0;
        if (dp[i] != -1) return dp[i];
        int ans = 1 + dfs(i + 1);
        
        for (int j = i + 1; j <= s.length(); j++) {
            if (set.contains(s.substring(i, j))) {
                ans = Math.min(ans, dfs(j));
            }
        }

        return dp[i] = ans;
    }
}

File: 2755-extra-characters-in-a-string/extra-characters-in-a-string.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWord = False
class Trie:
    def __init__(self):
        self.node = TrieNode()
    def addWord(self,word):
        currNode = self.node
        for i,ch in enumerate(word):
            if ch not in currNode.children:
                currNode.children[ch] = TrieNode()
            if i == len(word)-1:
                currNode.children[ch].isWord = True 
            currNode = currNode.children[ch]
    def searchWord(self,word):
        currNode = self.node
        for i,ch in enumerate(word):
            if ch not in currNode.children:
                return False
            if i == len(word)-1 and currNode.children[ch].isWord == False:
                return False
            currNode = currNode.children[ch]
        return True
            
            

class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        trie = Trie()
        for word in dictionary:
            trie.addWord(word)
        dp = {}
        def REC(index):
            if index == len(s):
                return 0
            if index in dp:
                return dp[index]
            chars = 1 + REC(index + 1)
            currNode = trie.node
            for i in range(index, len(s)):
                ch = s[i]
                if ch in currNode.children:
                    if currNode.children[ch].isWord == True:
                        chars = min(chars,REC(i+1))
                    currNode = currNode.children[ch]
                elif ch not in currNode.children:
                    break
            dp[index] = chars
            return chars
        return REC(0)
